### ISSUE TITLE
Implementation of `printable`

### ISSUE THREAD

**Issue by sticilface** [bblanchon/ArduinoJson#727]
Hi there, 

I've recently been playing with 

https://github.com/arduino/Arduino/blob/master/hardware/arduino/avr/cores/arduino/Printable.h

I was wondering if it might be possible to add such a feature to arduinoJson to allow such features as

```cpp
Serial.println(JsonObject &) 
```

It would require JsonObject, JsonArray, JsonVariant inheriting from `printable` and defining a virtual function `printTo`.  

I suspect the following might stop you.  Code size increase from adding virtual function + inherited class to each object + what would you print by default... pretty or standard... 

I just thought I would ask as there are no issues on printable... 

Cheers

A

**Comment by bblanchon** [comment]
Hi @sticilface,

Early versions of ArduinoJson implemented the `Printable` interface, but [I removed it in version 4.0](https://blog.benoitblanchon.fr/arduino-json-v4-0/) because it made `JsonArray`, `JsonObject` and `JsonVariant` bigger for little benefit.

However, it's possible that this feature will make a come back when I solve #309.
I don't know when this will happen though, because [version 6](https://github.com/bblanchon/ArduinoJson/tree/6.x) is all about the new serialization API and MsgPack.

Regards,
Benoit

**Comment by sticilface** [comment]
Hi,  thanks for the quick reply!

That makes sense if you can bundle it with a wrapper class.   I was curious more than anything.  not a super urgent feature but it would make printing it to certain things a little easier!

Looking forward to msg pack!!!

**Comment by bblanchon** [comment]
Hi @sticilface,

Bad news 😢 

I tried to implement `Printable` in the proxy classes of ArduinoJson v6, but the impact on the code is huge.
It tested on ATmega328; as soon as I inherit from `Printable`, the code grows by 1762 bytes, even if the program doesn't use this feature.

I looked at the output from `avr-nm`, and I saw that the compiler includes the JSON serializer, even if you don't use it in your program.
It means that if you only use MsgPack or only deserialize JSON, you pay the price of the serializer.

I tried to make the class and the function as `final`, and to inline most functions, but to no avail.
The compiler is not able to eliminate the extra weight.

So, I guess we have to forget about this feature.
Too bad the original Arduino developers didn't use a template overload of `Print::print()` instead of an abstract class.

Regards,
Benoit
